### ISSUE TITLE
Microsoft Office 365 timezones

### ISSUE THREAD

**Issue by jasonheffner** [u01jmg3/ics-parser#178]
> :information_source:
> - Firstly, check you are using the latest version (`dev-master`) as the problem may already have been fixed.
> - It is **essential** to be provided with the offending iCal causing the parser to behave incorrectly.
>   - Best to upload the iCal file directly to this issue

- PHP Version: `7.2.5`
- PHP date.timezone: `America/New_York`
- ICS Parser Version: `dev-master`
- Windows/Mac/Linux

I've had some problems parsing Microsoft ical files, in particular, the timezone naming formats used by Microsoft. I'm able to do a replacement after ics download and have been handling that this way.

i.e. - Replace "(UTC-05:00) Eastern Time (US & Canada)" with America/New_York

Here is a listing of the Microsoft TZ naming conventions used within the shared ics file urls..

http://unicode.org/repos/cldr/trunk/common/supplemental/windowsZones.xml
https://github.com/nehero/friendlytimezone

perhaps you have a different way of handling Microsoft TimeZones?

**Comment by u01jmg3** [comment]
Can you provide an unaltered example of the offending iCal (with the Microsoft CLDR time zone)?

**Comment by jasonheffner** [comment]
Here is a short example showing the Office365 layout along with TimeZones.

BEGIN:VCALENDAR
METHOD:PUBLISH
PRODID:Microsoft Exchange Server 2010
VERSION:2.0
X-WR-CALNAME:Calendar
BEGIN:VTIMEZONE
TZID:(UTC-05:00) Eastern Time (US & Canada)
BEGIN:STANDARD
DTSTART:16010101T020000
TZOFFSETFROM:-0400
TZOFFSETTO:-0500
RRULE:FREQ=YEARLY;INTERVAL=1;BYDAY=1SU;BYMONTH=11
END:STANDARD
BEGIN:DAYLIGHT
DTSTART:16010101T020000
TZOFFSETFROM:-0500
TZOFFSETTO:-0400
RRULE:FREQ=YEARLY;INTERVAL=1;BYDAY=2SU;BYMONTH=3
END:DAYLIGHT
END:VTIMEZONE
BEGIN:VEVENT
UID:040000008200E00074C5B7101A82E008000000004A3BC21C0FA4D301000000000000000
 01000000020F1C02EB6C39045805852176FE471A7
SUMMARY:Canvas Day Discussion
DTSTART;TZID="(UTC-05:00) Eastern Time (US & Canada)":20180215T103000
DTEND;TZID="(UTC-05:00) Eastern Time (US & Canada)":20180215T110000
CLASS:PUBLIC
PRIORITY:5
DTSTAMP:20180515T123810Z
TRANSP:OPAQUE
STATUS:CONFIRMED
SEQUENCE:1
LOCATION:Meeting Room 022e
X-MICROSOFT-CDO-APPT-SEQUENCE:1
X-MICROSOFT-CDO-BUSYSTATUS:BUSY
X-MICROSOFT-CDO-INTENDEDSTATUS:BUSY
X-MICROSOFT-CDO-ALLDAYEVENT:FALSE
X-MICROSOFT-CDO-IMPORTANCE:1
X-MICROSOFT-CDO-INSTTYPE:0
X-MICROSOFT-DONOTFORWARDMEETING:FALSE
X-MICROSOFT-DISALLOW-COUNTER:FALSE
END:VEVENT
END:VCALENDAR

**Comment by u01jmg3** [comment]
Thanks - what output do you expect to see? What output are you getting instead?

**Comment by u01jmg3** [comment]
- Using `dev-master` and the iCal you provided I am seeing no such error.
- What exact code are you using with the parser to read your iCal?
- What version of Carbon are you using?

**Comment by jasonheffner** [comment]
using dev-master .. here is composer output 

johngrogg/ics-parser      dev-master 19931c8 ICS Parser
nesbot/carbon             1.27.0             A simple API extension for DateTime.
symfony/polyfill-mbstring v1.8.0             Symfony polyfill for the Mbstring extension
symfony/translation       v4.0.9             Symfony Translation Component

I started seeing this when we transitioned to Office 365 with "Eastern Standard Time" and "Eastern Daylight Time" so simply did a sed replacement after importing. I just recently started seeing "(UTC-05:00) Eastern Time (US & Canada)" timezones in the ical feeds from Office 365. I was assuming that they were unknown to your parser. 

I use the following code

use ICal\ICal;
$ical = new ICal('data.txt', array(
    'defaultSpan'           => 2,     // Default value
    'defaultWeekStart'      => 'MO',  // Default value
    'skipRecurrence'        => false, // Default value
    'useTimeZoneWithRRules' => false, // Default value
));

$events = $ical->eventsFromRange($dtStartDay, $dtEndDay);

**Comment by jasonheffner** [comment]
btw .. "Eastern Standard Time" and "Eastern Daylight Time" are deprecated TimeZone strings hence why I did the replacement. I think MS caught on and hence why I'm seeing the new timezone strings on some of their feeds, though without consistency.

**Comment by u01jmg3** [comment]
According to the error output you provided, the iCal should contain exdates which your snippet doesn't hence why I wasn't seeing the error you were. This is why it is essential I am provided with the offending iCal.

Anyway I was able to come up with my own iCal that caused the same error you were seeing (although I prefer not to have to do this in case I miss or overlook something).

Checkout my new branch called [`feature-cldr-time-zones`](https://github.com/u01jmg3/ics-parser/tree/feature-cldr-time-zones) which should provide basic support for CLDR time zones although I'd want you to throw more Microsoft exported iCal at this version of the parser to ensure it works under multiple scenarios.

**Comment by jasonheffner** [comment]
Sorry I couldn't share the ics file publically. I can share it privately. I understand needing the test case. 

I was able to get libicu updated and php-intl installed to support IntlTimeZone::getIDForWindowsID. I'd imagine the PHP ≥ 7.1 and ICU version ≥ 52 requirements could be a concern. It was a bit tricky of an install on CentOS 7 and could cause a lot of questions. Perhaps it was an easy install on your setup.

That said, it parsed all 5 of my daily ical feeds with the new TZ strings, including the older deprecated TZ strings. The code looks clean too. Quick work. The parser seems to be doing well on all the calendar feeds. I'm running it for room signage so errors get noticed quickly. I only notice the rare occasional duplicate usually due to exdates, which I have been parsing out manually.

If you need more help testing versions let me know.

**Comment by u01jmg3** [comment]
[According to the manual](https://secure.php.net/manual/en/intl.installation.php) the `php_intl` extension should be available on PHP 5.3.0+.

Nonetheless, time zones rarely change ([14 Jul 2009](https://support.microsoft.com/en-gb/help/973627/microsoft-time-zone-index-values)) so I have prerun the conversion - see this [new branch](https://github.com/u01jmg3/ics-parser/compare/master...feature-cldr-time-zones-no-intl) which does not need the `php_intl` extension.

---

> _I only notice the rare occasional duplicate usually due to exdates, which I have been parsing out manually._

With more info could I fix this?

**Comment by jasonheffner** [comment]
Unfortunately, certain classes and functions have higher requirements. It's not all that clear on the manual and is really annoying as you have to check each function. The IntlTimeZone class requires requires PHP 5 ≥ 5.5.0 and the function [getIDForWindowsID](http://php.net/manual/en/intltimezone.getidforwindowsid.php) requires PHP 7 ≥ 7.1.0 and ICU version ≥ 52. The latter is an issue on most distros as the libicu package is much older and all the packages don't include that function. I had to compile libicu seperately and then php_intl to make it work.

Perfect, removing the php_intl requirement will cause less headaches. I deployed that branch earlier this morning and looking over the code I don't see any issues.

I shared a private repo with a sample code and ics file for the duplicate.

**Comment by u01jmg3** [comment]
> _I shared a private repo with a sample code and ics file for the duplicate._

- The issue is non-unique UIDs in the private iCal snippet.

As per the spec:
[The UID itself MUST be a globally unique identifier. The generator of the identifier MUST guarantee that the identifier is unique.](https://www.kanzaki.com/docs/ical/uid.html)

- I guess Office 365 is not conforming again?
- Not sure what to do about this one...

**Comment by jasonheffner** [comment]
I double-checked the UID in the ics file I provided and the UID is unique for that meeting. Is the problem the UID is so long that it spans to the next row? I uploaded a second ics with just the faulty meeting in it.

**Comment by jasonheffner** [comment]
I went further through the ics file and code and found that it may be occurring due to the unimplemented BYDAY RRule. Every RECURRENCE-ID exception to the RRule gets duped.

**Comment by jasonheffner** [comment]
I found that the issue arises when there is an exception on the very first date of a recurring meeting.

**Comment by u01jmg3** [comment]
> _I double-checked the UID in the ics file I provided and the UID is unique for that meeting. There are 10 occurrences in the file._

Should the UID not be different for each of the 10 occurrences (even if they have the same summary)? I would expect to see each new `BEGIN:VEVENT...END:VEVENT` to have a new UID. If an event relates to another then the [`RELATED-TO`](https://www.kanzaki.com/docs/ical/relatedTo.html) attribute should be used.

>_Is the problem the UID is so long that it spans to the next row?_

No, [unfolding](https://github.com/u01jmg3/ics-parser/blob/master/src/ICal/ICal.php#L427-L441) is taken care of.

**Comment by jasonheffner** [comment]
That's a good question, I don't know. I assumed the same UID since it's for the same meeting using the RECCURENCE-ID and SEQUENCE number. I'll compare some other meetings and ics files from the old system and get back in touch tomorrow. I know I had the same issue before the transition to O365 and why I wrote the filter to remove any dupes there as well.

**Comment by jasonheffner** [comment]
I checked Zimbra and Office 365 and they are the same. I also looked at the spec here 

https://www.kanzaki.com/docs/ical/recurrenceId.html

The UID is unique per meeting, however, RECCURENCE-IDs, or rather exceptions in the recurrences, use RECURRENCE-ID in conjunction with the "UID" and "SEQUENCE" property to identify a specific instance according to the spec. This coincides with what I'm seeing in the ics files. Perhaps a new issue?

**Comment by u01jmg3** [comment]
a9fc44006062522a870 should do the trick

**Comment by jasonheffner** [comment]
Yep, that seems to have done it. I tested on the two examples I had and can verify more tomorrow. The event count wasn't quite right and it threw some warnings, so I made a quick edit and left a comment on your merge. Hopefully, I made the right assumptions.

**Comment by jasonheffner** [comment]
I updated my comment and must have been tired last night. There seems to be a small bug but I'm guessing easily fixed. In my case the example I used the first event in the series is also altered.

**Comment by u01jmg3** [comment]
Corrected in b15e79b. This should also fix the count being out by `1`.

Good to go?

**Comment by jasonheffner** [comment]
Fixed the dupes and the warning.. but count still off by 1. I'm assuming it's the occasion when the initial first occurrence also has an exception.

**Comment by u01jmg3** [comment]
- Can you confirm expected and actual counts for me to work with?
- Is this for `data22e.ics`?

**Comment by jasonheffner** [comment]
This is for tech_meeting.ics in the git repo. You can look at example.php and example.html. I tested by removing the first RECURRENCE-ID, for the first start time, then removing other RECURRENCE-IDs. The count is correct for all but the initial test case.

**Comment by u01jmg3** [comment]
To confirm you get `37` but you expect `38`?

**Comment by jasonheffner** [comment]
I expect to see 36 from the listing but get 37. I removed the first exception that happens on the first occurrence and it's 36. Any other change doesn't affect the count. I was going to dump the array next to see what is going on with maybe only 2 recurrence-ids

**Comment by jasonheffner** [comment]
Now I see how you organize the array .. have 36(0-35) events stored in the array, but eventCount is at 37. Must have missed a decrement somewhere.

**Comment by jasonheffner** [comment]
I may have found a fix .. add this to line [888](https://github.com/u01jmg3/ics-parser/blob/feature-cldr-time-zones-no-intl/src/ICal/ICal.php#L888)

$this->eventCount = $this->eventCount - count($eventKeysToRemove);

I'd verify but appears that this will work .. short of using 

$this->eventCount = count($this->cal['VEVENT']);

to track array size. It's a decent debug check if the two coincide after parsing an ical file.

**Comment by u01jmg3** [comment]
See 72d3d0c - all good?

**Comment by jasonheffner** [comment]
That fixes the count correctly, however, I was running through all my test files and found an issue with the original commit [a9fc440](https://github.com/u01jmg3/ics-parser/commit/a9fc44006062522a8700ecb2d4ee20d4e4f0ffe9). 

I had been testing with the shortened version of the ical file only containing that one tech meeting. This works perfectly now. When using the entire calendar export with multiple meetings intertwined I see duplicates for the first meeting in the series only. The entries are back to back in the file. I haven't tracked down the cause yet and perhaps you have an idea? It's only the first occurrence, works as expected in the shortened ical, but duplicates when parsing the entire ical export. Perhaps a variable not properly cleared on loop?

**Comment by u01jmg3** [comment]
Hopefully my final commit: a73c6b4

You'll notice the code I added previously for fixing the event count is no longer necessary and has been removed.

Please retest against all your iCal snippets paying attention to the event count also.

Many thanks

**Comment by jasonheffner** [comment]
Yes, I think it's your last commit for this issue. I ran it through all my test files, including 5 more rooms, and works perfectly. I checked for dupes, and compared $ical->eventCount to count($ical->cal['VEVENT']) to check for any inconsistencies and they all match. 

When I release this to our code on Monday I'll write a sanity check in for comparing the counts and if they don't match save the offending ical to a new folder and notify me. The parser runs every 5 minutes checking for changes on several rooms. I hope to deploy the signage to more than our one test location eventually.

Many thanks to you as well!
